src/lib/sort/sort-header.ts

```diff
@@ -50,52 +60,135 @@ const SORT_ANIMATION_TRANSITION =
-      transition('void => asc', animate(SORT_ANIMATION_TRANSITION, keyframes([
-        style({transform: 'translateY(25%)', opacity: 0}),
-        style({transform: 'none', opacity: 1})
+
+    // Translation of the arrow as a whole. States are separated into two groups: ones with
+    // animations and others that are immediate. Immediate states are asc, desc, peek, and active.
+    // The other states define a specific animation (source-to-destination) and are determined as
+    // a function of their prev user-perceived state and what the next state should be.
+    trigger('arrowPosition', [
+      // Hidden Above => Peek Center
+      transition('* => desc-to-peek, * => desc-to-active',
+          animate(SORT_ANIMATION_TRANSITION, keyframes([
+            style({transform: 'translateY(-25%)'}),
+            style({transform: 'translateY(0)'})
       ]))),
-      transition('asc => void', animate(SORT_ANIMATION_TRANSITION, keyframes([
-        style({transform: 'none', opacity: 1}),
-        style({transform: 'translateY(-25%)', opacity: 0})
+      // Peek Center => Hidden Below
+      transition('* => peek-to-desc, * => active-to-desc',
+          animate(SORT_ANIMATION_TRANSITION, keyframes([
+            style({transform: 'translateY(0)'}),
+            style({transform: 'translateY(25%)'})
       ]))),
-      transition('void => desc', animate(SORT_ANIMATION_TRANSITION, keyframes([
-        style({transform: 'translateY(-25%)', opacity: 0}),
-        style({transform: 'none', opacity: 1})
+      // Hidden Below => Peek Center
+      transition('* => asc-to-peek, * => asc-to-active',
+          animate(SORT_ANIMATION_TRANSITION, keyframes([
+            style({transform: 'translateY(25%)'}),
+            style({transform: 'translateY(0)'})
       ]))),
-      transition('desc => void', animate(SORT_ANIMATION_TRANSITION, keyframes([
-        style({transform: 'none', opacity: 1}),
-        style({transform: 'translateY(25%)', opacity: 0})
+      // Peek Center => Hidden Above
+      transition('* => peek-to-asc, * => active-to-asc',
+          animate(SORT_ANIMATION_TRANSITION, keyframes([
+            style({transform: 'translateY(0)'}),
+            style({transform: 'translateY(-25%)'})
       ]))),
+      state('desc-to-peek, asc-to-peek, peek, desc-to-active, asc-to-active, active',
+          style({transform: 'translateY(0)'})),
+      state('peek-to-desc, active-to-desc, desc',
+          style({transform: 'translateY(-25%)'})),
+      state('peek-to-asc, active-to-asc, asc',
+          style({transform: 'translateY(25%)'})),
     ])
   ]
 })
 export class MatSortHeader implements MatSortable {
   private _rerenderSubscription: Subscription;
 
+  /**
+   * Flag set to true when the indicator should be displayed while the sort is not active. Used to
+   * provide an affordance that the header is sortable by showing on focus and hover.
+   * @docs-private
+   */
+  set showIndicatorHint(showIndicatorHint: boolean) {
```

Comment:
Instead of making this a setter, can you make it a function like `_setIndicatorHintVisible`?